docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,228 @@
+| `setmetatable(arg: typelib)` | `nil` | sets self's metatable to the argument; both self and the argument need to be `ismetatable()` |
+| `getmetatable()` | `typelib?` | returns self's runtime representation of metatable if it exists, else nil; self needs to be `ismetatable()` |
+
+#### Function
+
+| Instance Methods | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `setparameters(arg: {typelib} \| typelib?)` | `nil` | sets self's parameter types to the argument, where an array implies a TypePack and the latter implies a Variadic |
+| `getparameters()` | `{typelib} \| typelib?` | returns the runtime representation of self's parameter type if it exists, else nil. Return an array implies a TypePack and a single value implies a Variadic |
+| `setreturns(arg: {typelib} \| typelib?)` | `nil` | sets self's return types to the argument, where an array implies a TypePack and the latter implies a Variadic |
+| `getreturns()` | `{typelib} \| typelib?` | returns the runtime representation of self's return type if it exists, else nil. Return an array implies a TypePack and a single value implies a Variadic |
+
+#### Union
+
+| Instance Methods | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getcomponents()` | `{typelib}` | returns an array of types that the self's union can represent. For instance, `string \| number` returns `{typelib.getstring(), typelib.getnumber()}` |
+
+#### Intersection
+
+| Instance Methods | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getcomponents()` | `{typelib}` | returns an array of types represented by self's intersection. For instance, `string & number` returns `{typelib.getstring(), typelib.getnumber()}` |
+
+#### Class
+
+| Instance Methods | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getprops()` | `{[typelib]: typelib}` | returns the runtime representation self's properties |
+| `getparent()` | `typelib?` | returns the runtime representation of self's parent class if it exists, else nil |
+| `getmetatable()` | `typelib?` | returns the runtime representation of self's metatable if it exists, else nil |
+| `getindexer()` | `{key: typelib, value: typelib}?` | returns a table containing self's indexer key type and value type |
+
+</details>
+
+The reason for going with userdata instead using another representation or adding a new compile-time interpreter like many other language is that userdata provides a clean abstraction from the runtime representation and restricts developers to using a set of libraries controlled by the Luau team where we can easily manage changes and listen to feedbacks. Moreover, because Lua was designed to be easily embeddable with C++, we wanted to use this as an advantage for type functions since they only require using a small runtime variant of the VM, making the implementation be simpler than having to implement a new interpreter.
```

Comment:
```suggestion
The reason for going with userdata instead using another representation or adding a new compile-time interpreter like many other language is that userdata provides a clean abstraction from the runtime representation and restricts developers to using a set of libraries controlled by the Luau team where we can easily manage changes and listen to feedbacks. Moreover, because Luau is designed to be easily embeddable with C++, we wanted to use this as an advantage for type functions since they only require using a small runtime variant of the VM, making the implementation be simpler than having to implement a new interpreter.
```